### PR TITLE
Parcelport plugins

### PR DESCRIPTION
**Pull request by sithhell**
The main addition of this PR is the ability to have parcelport plugins (both statically compiled into the main HPX library and dynamically loadable). This addresses #1201 and #680. In addition, the MPI parcelport has been rewritten in such a way that it reaches an acceptable performance (#1275). The IPC and ibverbs parcelports are temporarly disabled since they are non functional and need to be fixed. The changeset also includes various other code improvements.

### CODE REVIEW

File: hpx/hpx_fwd.hpp [hunk 1]
```diff
@@ -1453,39 +1437,6 @@ namespace hpx
...
-    ///
-    /// \returns  This function returns the name for the locality of the given
-    ///           id. The name is retrieved from the underlying networking layer
-    ///           and may be different for different parcel ports.
-    ///
-    /// \see      \a std::string get_locality_name()
-    HPX_API_EXPORT future<std::string> get_locality_name(
-        naming::id_type const& id);
-
-    ///////////////////////////////////////////////////////////////////////////
```
Comment (hkaiser) [inline hpx/hpx_fwd.hpp]: Are these API functions to be removed?

File: hpx/lcos/future.hpp [hunk 2]
```diff
@@ -1117,17 +1117,25 @@ namespace hpx { namespace lcos
...
+            typedef
+                typename traits::future_traits<shared_future>::result_type
+                result_type;
             if (!this->shared_state_)
             {
                 HPX_THROWS_IF(ec, no_state,
                     "shared_future<R>::get",
                     "this future has no valid shared state");
-                return detail::future_value<R>::get_default();
+                static result_type tmp(detail::future_value<R>::get_default());
```
Comment (hkaiser) [inline hpx/lcos/future.hpp]: What is this change all about? I don't think this is correct.

File: hpx/lcos/future.hpp [hunk 3]
```diff
@@ -1117,17 +1117,25 @@ namespace hpx { namespace lcos
...
+                return tmp;
             }
 
             typedef typename shared_state_type::data_type data_type;
             data_type& data = this->shared_state_->get_result(ec);
-            if (ec) return detail::future_value<R>::get_default();
+            if (ec)
+            {
+                static result_type tmp(detail::future_value<R>::get_default());
```
Comment (hkaiser) [inline hpx/lcos/future.hpp]: Same as above, this does not look right.

File: hpx/lcos/local/recursive_mutex.hpp [hunk 4]
```diff
@@ -176,6 +180,8 @@ namespace hpx { namespace lcos { namespace local
             {
                 if (mtx.try_lock())
                 {
+                    util::ignore_lock(&mtx);
+                    //util::register_lock(this);
```
Comment (hkaiser) [inline hpx/lcos/local/recursive_mutex.hpp]: Where is the `ignore_lock` reset?

File: hpx/hpx_fwd.hpp (hunk 1 above)
Comment (sithhell) [inline hpx/hpx_fwd.hpp]: Those are not used anywhere and locality name doesn't make lots of sense without a defined parcelport.

File: hpx/plugins/parcelport/mpi/mpi_environment.hpp [hunk 5]
```diff
@@ -29,27 +30,38 @@ namespace hpx { namespace util
...
+        };
+
+        struct scoped_try_lock
+        {
+            scoped_try_lock();
+            ~scoped_try_lock();
+            bool locked;
+        };
+
+        typedef hpx::lcos::local::spinlock mutex_type;
```
Comment (hkaiser) [inline hpx/plugins/parcelport/mpi/mpi_environment.hpp]: Why do you need your own lock here, just use `unique_lock`. Also using a non-atomic `bool` looks suspicious here.

File: hpx/lcos/future.hpp (hunk 2 above)
Comment (sithhell) [inline hpx/lcos/future.hpp]: This is about sliencing the warning about returning a reference to a temporary. `shared_future<T>::get(error_code)` return a const reference.

File: hpx/plugins/parcelport/mpi/mpi_environment.hpp [hunk 6]
```diff
@@ -29,27 +30,38 @@ namespace hpx { namespace util
...
 
-#else
+        static mutex_type mtx_;
 
-namespace hpx { namespace util
-{
-    struct command_line_handling;
+        static void lock();
+        static bool try_lock();
+        static void unlock();
```
Comment (hkaiser) [inline hpx/plugins/parcelport/mpi/mpi_environment.hpp]: These don't look right. Why do you need your own locking here?

File: hpx/lcos/local/recursive_mutex.hpp (hunk 4 above)
Comment (sithhell) [inline hpx/lcos/local/recursive_mutex.hpp]: When the lock is unlocked again, it's removed from the held locks map and therefor reset.

File: hpx/plugins/parcelport_factory.hpp [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+/// This macro is used create and to register a minimal component factory with
+/// Hpx.Plugin.
+#define HPX_REGISTER_PARCELPORT_(Parcelport, pluginname, pp)                  \
+    template struct hpx::plugins::parcelport_factory<Parcelport>;             \
+    typedef hpx::plugins::parcelport_factory<Parcelport>                      \
+        BOOST_PP_CAT(pluginname, _plugin_factory_type);                       \
+    static BOOST_PP_CAT(pluginname, _plugin_factory_type)                     \
+        BOOST_PP_CAT(pluginname, _factory);                                   \
+    HPX_EXPORT hpx::plugins::parcelport_factory_base * BOOST_PP_CAT(pluginname, _factory_base_ptr) = &BOOST_PP_CAT(pluginname, _factory); \
+    HPX_DEF_UNIQUE_PLUGIN_NAME(                                               \
```
Comment (hkaiser) [inline hpx/plugins/parcelport_factory.hpp]: Please make this line 80 characters wide at max.

File: hpx/plugins/parcelport_factory_base.hpp [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+/*
+///////////////////////////////////////////////////////////////////////////////
+/// This macro is used to register the given component factory with
+/// Hpx.Plugin. This macro has to be used for each of the component factories.
+#define HPX_REGISTER_PARCELPORT_BASE(FactoryType, pluginname)                 \
+    HPX_PLUGIN_EXPORT(HPX_PLUGIN_PLUGIN_PREFIX,                               \
+        hpx::plugins::plugin_factory_base, FactoryType,                       \
+        pluginname, factory)                                                  \
+*/
+/**/
```
Comment (hkaiser) [inline hpx/plugins/parcelport_factory_base.hpp]: Can this be removed?

File: hpx/plugins/parcelport_factory_base.hpp [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+        ///
+        /// return Returns the newly created instance of the parcelport
+        ///        supported by this factory
+        virtual parcelset::parcelport* create(
+            hpx::util::runtime_configuration const & cfg,
+            HPX_STD_FUNCTION<void(std::size_t, char const*)> const& on_start_thread,
+            HPX_STD_FUNCTION<void()> const& on_stop_thread) = 0;
+    };
+}}
+
```
Comment (hkaiser) [inline hpx/plugins/parcelport_factory_base.hpp]: Don't you need a specialization of  

    template <>                                                          
    struct virtual_constructors<parcelport_factory_base> 

here?

File: hpx/runtime/agas/big_boot_barrier.hpp [hunk 10]
```diff
@@ -67,7 +67,7 @@ struct HPX_EXPORT big_boot_barrier : boost::noncopyable
     };
 
     big_boot_barrier(
-        parcelset::parcelport& pp_
+        parcelset::parcelport* pp_
       , parcelset::endpoints_type const& endpoints_
```
Comment (hkaiser) [inline hpx/runtime/agas/big_boot_barrier.hpp]: Why did that change to a pointer here?

File: hpx/lcos/future.hpp (hunk 2 above)
Comment (K-ballo) [inline hpx/lcos/future.hpp]: Wouldn't it be cleaner to just make `get_default` return a const reference to static?

Comment (sithhell) [inline hpx/lcos/future.hpp]: Yes, now that you mention it ...

File: hpx/plugins/parcelport/mpi/mpi_environment.hpp (hunk 6 above)
Comment (sithhell) [inline hpx/plugins/parcelport/mpi/mpi_environment.hpp]: The look is used to protect the calls to MPI. Locking gets automatically disabled if we have a multithreaded MPI at hand.

File: hpx/plugins/parcelport_factory.hpp (hunk 7 above)
Comment (sithhell) [inline hpx/plugins/parcelport_factory.hpp]: Thanks, forgot to reformat that.

File: hpx/plugins/parcelport_factory_base.hpp (hunk 8 above)
Comment (sithhell) [inline hpx/plugins/parcelport_factory_base.hpp]: Yes.

File: hpx/plugins/parcelport_factory_base.hpp (hunk 9 above)
Comment (sithhell) [inline hpx/plugins/parcelport_factory_base.hpp]: Nope. This works. `parcelport_factory_base` is not derived from the plugin_base.

File: hpx/runtime/agas/big_boot_barrier.hpp (hunk 10 above)
Comment (sithhell) [inline hpx/runtime/agas/big_boot_barrier.hpp]: Because it might be null if no parcelport was configured.

File: hpx/plugins/parcelport/mpi/mpi_environment.hpp (hunk 5 above)
Comment (sithhell) [inline hpx/plugins/parcelport/mpi/mpi_environment.hpp]: The non atomic is bool is fine. There should only be a single reference to a object of scoped_lock anyways. Probably should make that non copyable and non movable. (compare to boost::unique_lock).

File: hpx/lcos/future.hpp (hunk 2 above)
Comment (K-ballo) [inline hpx/lcos/future.hpp]: Actually, no. That would create problems with `future<R>` for movable-only `R`s. But at least give it a prettier name than `tmp`.

File: hpx/util/asio_util.hpp [hunk 11]
```diff
@@ -66,7 +66,7 @@ namespace hpx { namespace util
         return accept_begin(loc.address(), loc.port(), io_service);
     }
 
-    inline endpoint_iterator_type accept_end() //-V524
+    inline endpoint_iterator_type HPX_EXPORT accept_end()
     {
```
Comment (hkaiser) [inline hpx/util/asio_util.hpp]: inline functions don't have to HPX_EXPORT

File: hpx/util/register_locks.hpp [hunk 12]
```diff
@@ -127,9 +127,6 @@ namespace hpx { namespace util
     inline void force_error_on_lock()
     {
     }
-    inline void enable_lock_detection()
-    {
-    }
     inline void ignore_lock(void const* /*lock*/)
```
Comment (hkaiser) [inline hpx/util/register_locks.hpp]: Why was this removed

File: plugins/parcel/CMakeLists.txt [hunk 13]
```diff
@@ -3,8 +3,8 @@
 # Distributed under the Boost Software License, Version 1.0. (See accompanying
 # file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
 
-set(parcel_plugins
-    coalescing)
+set(parcel_plugins)
+    #coalescing)
```
Comment (hkaiser) [inline plugins/parcel/CMakeLists.txt]: Why has this been commented out?

File: plugins/parcelport/ibverbs/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+              "${hpx_SOURCE_DIR}/hpx/plugins/parcelport/ibverbs/receiver.hpp"
+              "${hpx_SOURCE_DIR}/hpx/plugins/parcelport/ibverbs/sender.hpp"
+              "${hpx_SOURCE_DIR}/hpx/plugins/parcelport/ibverbs/server.hpp"
+        DEPENDENCIES
+              ${IBVERBS_LIBRARIES}
+              ${RDMACM_LIBRARIES}
+        FOLDER "Core/Plugins/Parcelport/Ibverbs"
+        )
+  endmacro()
+else()
```
Comment (hkaiser) [inline plugins/parcelport/ibverbs/CMakeLists.txt]: Can't this use globbing (as add_component() et.al.) instead of explicitly listing the files?

File: plugins/parcelport/mpi/CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+        "${hpx_SOURCE_DIR}/hpx/plugins/parcelport/mpi/sender.hpp"
+        "${hpx_SOURCE_DIR}/hpx/plugins/parcelport/mpi/receiver.hpp"
+      DEPENDENCIES
+        ${MPI_CXX_LIBRARIES}
+        ${MPI_LIBRARY}
+        ${MPI_EXTRA_LIBRARY}
+      FOLDER "Core/Plugins/Parcelport/MPI"
+    )
+  endmacro()
+else()
```
Comment (hkaiser) [inline plugins/parcelport/mpi/CMakeLists.txt]: Same as above, use GLOB

File: plugins/parcelport/mpi/mpi_environment.cpp [hunk 16]
```diff
@@ -137,15 +136,20 @@ namespace hpx { namespace util
...
+        }
 
+        if(provided_threading_flag_ == MPI_THREAD_FUNNELED)
+        {
             enabled_ = false;
             has_called_init_ = false;
             throw std::runtime_error("mpi_environment::init: MPI_Init_thread: "
-                "provided multi_threading mode is different from requested mode");
+                "The underlying MPI implementation only supports MPI_THREAD_FUNNELED. This mode is not supported by HPX. Please pass -Ihpx.parcel.mpi.multithreaded=0 to explicitly disable MPI multithreading.");
         }
```
Comment (hkaiser) [inline plugins/parcelport/mpi/mpi_environment.cpp]: Please keep line length at around 80 characters at most

File: hpx/util/register_locks.hpp (hunk 12 above)
Comment (sithhell) [inline hpx/util/register_locks.hpp]: It's not needed in the case that lock detection is disabled at compile time. Why keep dead code?

File: plugins/parcel/CMakeLists.txt (hunk 13 above)
Comment (sithhell) [inline plugins/parcel/CMakeLists.txt]: My mistake. Will reenable it.

File: src/runtime/components/server/runtime_support_server.cpp [hunk 17]
```diff
@@ -966,7 +967,7 @@ namespace hpx { namespace components { namespace server
         }
 
         plugins_.clear();       // unload all plugins
-        modules_.clear();       // unload all modules
+        //modules_.clear();       // unload all modules
     }
```
Comment (hkaiser) [inline src/runtime/components/server/runtime_support_server.cpp]: Can this be removed?

File: plugins/parcelport/ibverbs/CMakeLists.txt (hunk 14 above)
Comment (sithhell) [inline plugins/parcelport/ibverbs/CMakeLists.txt]: Could, yes. I just didn't implement that for this function.

File: src/runtime_impl.cpp [hunk 18]
```diff
@@ -145,7 +141,7 @@ namespace hpx {
         thread_manager_(new hpx::threads::threadmanager_impl<
             SchedulingPolicy, NotificationPolicy>(
                 timer_pool_, scheduler_, notifier_, num_threads)),
-        parcel_handler_(thread_manager_.get(),
+        parcel_handler_(rtcfg, agas_client_, thread_manager_.get(),
             new parcelset::policies::global_parcelhandler_queue,
```
Comment (hkaiser) [inline src/runtime_impl.cpp]: This uses `agas_client_` before it is initialized.

Comment (hkaiser) [inline src/runtime_impl.cpp]: I believe this is the result of a bad merge as I remember to have fixed that already.

File: hpx/hpx_fwd.hpp (hunk 1 above)
Comment (hkaiser) [inline hpx/hpx_fwd.hpp]: This is part of the API and might not be used by HPX itself, but even MPI has `MPI_Get_processor_name`, thus people expect to have. At minimum it should return "locality#N". But I believe the TCP pp will be available almost everywhere, thus the hostname should be available.

File: hpx/lcos/local/recursive_mutex.hpp (hunk 4 above)
Comment (hkaiser) [inline hpx/lcos/local/recursive_mutex.hpp]: AFAICS, you have it commented out: https://github.com/STEllAR-GROUP/hpx/blob/parcelport_plugins/hpx/lcos/local/recursive_mutex.hpp#L158

File: hpx/plugins/parcelport/mpi/mpi_environment.hpp (hunk 5 above)
Comment (hkaiser) [inline hpx/plugins/parcelport/mpi/mpi_environment.hpp]: Still, why do you need your own lock here? Those global lock() and unlock() functions are just wrong.

File: hpx/plugins/parcelport/mpi/mpi_environment.hpp (hunk 6 above)
Comment (hkaiser) [inline hpx/plugins/parcelport/mpi/mpi_environment.hpp]: Why not use scoped_lock or unique_lock?

File: hpx/runtime/agas/big_boot_barrier.hpp (hunk 10 above)
Comment (hkaiser) [inline hpx/runtime/agas/big_boot_barrier.hpp]: I now think this was part of my change to resolve the circular dependency between the initialization of AGAS and the parcel handler.

File: hpx/util/register_locks.hpp (hunk 12 above)
Comment (hkaiser) [inline hpx/util/register_locks.hpp]: For ABI compatibility?

File: plugins/parcelport/ibverbs/CMakeLists.txt (hunk 14 above)
Comment (hkaiser) [inline plugins/parcelport/ibverbs/CMakeLists.txt]: I'd strongly suggest implementing it. The way you have it looks just too painful.

File: src/runtime_impl.cpp (hunk 18 above)
Comment (hkaiser) [inline src/runtime_impl.cpp]: Indeed: https://github.com/STEllAR-GROUP/hpx/blob/master/src/runtime_impl.cpp#L151

File: hpx/plugins/parcelport/mpi/mpi_environment.hpp (hunk 6 above)
Comment (sithhell) [inline hpx/plugins/parcelport/mpi/mpi_environment.hpp]: As discussed on IRC, the global lock/unlock functions go, but the scoped_lock functionality needs to stay.

File: plugins/parcelport/ibverbs/CMakeLists.txt (hunk 14 above)
Comment (biddisco) [inline plugins/parcelport/ibverbs/CMakeLists.txt]: Globbing is generally not recommended with cmake as builds are not automatically triggered if files are added. Using explicit naming is preferred.

Comment (hkaiser) [inline plugins/parcelport/ibverbs/CMakeLists.txt]: Did I mention that cmake is a PITA?

File: hpx/hpx_fwd.hpp [hunk 19]
```diff
@@ -1703,6 +1654,9 @@ namespace hpx
 #endif
 }
 
+// Including declarations of get_locality_name
+#include <hpx/runtime/get_locality_name.hpp>
+
```
Comment (hkaiser) [inline hpx/hpx_fwd.hpp]: Why is this a separate header (I think we should keep all of the API in `hpx_fwd.hpp`)? Also, this header is currently missing from the repository?

File: hpx/util/memory_chunk_pool.hpp [hunk 20]
```diff
@@ -106,19 +117,99 @@ namespace hpx { namespace util {
...
+            typedef const char* const_pointer;
+            typedef char& reference;
+            typedef const char& const_reference;
+            typedef char value_type;
+
+            template <typename U>
+            struct rebind
+            {
+                typedef memory_chunk_pool_allocator other;
+            };
```
Comment (hkaiser) [inline hpx/util/memory_chunk_pool.hpp]: This implementation of rebind is not conforming. it needs to return a new allocator type which allocates objects of type `U`. This rebind always returns an allocator which allocates `char`. In conforming implementations this will not compile.

**Comment by hkaiser** [comment]
This looks reasonably good. We should merge it to have regular testing.

**Comment by sithhell** [comment]
The unit tests look pretty fine now. Please verify the latest changes once more.
